pkg/scalers/dynatrace_scaler.go

```diff
@@ -0,0 +1,242 @@
+package scalers
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	neturl "net/url"
+	"strconv"
+	"strings"
+
+	"github.com/go-logr/logr"
+	v2 "k8s.io/api/autoscaling/v2"
+	"k8s.io/metrics/pkg/apis/external_metrics"
+
+	"github.com/kedacore/keda/v2/pkg/scalers/scalersconfig"
+	kedautil "github.com/kedacore/keda/v2/pkg/util"
+)
+
+const (
+	dynatraceMetricDataPointsAPI = "api/v2/metrics/query"
+)
+
+type dynatraceScaler struct {
+	metricType v2.MetricTargetType
+	metadata   *dynatraceMetadata
+	httpClient *http.Client
+	logger     logr.Logger
+}
+
+type dynatraceMetadata struct {
+	host                string
+	token               string
+	metricSelector      string
+	fromTimestamp       string
+	threshold           float64
+	activationThreshold float64
+	triggerIndex        int
+}
+
+// Model of relevant part of Dynatrace's Metric Data Points API Response
+// as per https://docs.dynatrace.com/docs/dynatrace-api/environment-api/metric-v2/get-data-points#definition--MetricData
+type dynatraceResponse struct {
+	Result []struct {
+		Data []struct {
+			Values []float64 `json:"values"`
+		} `json:"data"`
+	} `json:"response"`
+}
+
+func NewDynatraceScaler(config *scalersconfig.ScalerConfig) (Scaler, error) {
+	metricType, err := GetMetricTargetType(config)
+	if err != nil {
+		return nil, fmt.Errorf("error getting scaler metric type: %w", err)
+	}
+
+	logger := InitializeLogger(config, "dynatrace_scaler")
+
+	meta, err := parseDynatraceMetadata(config, logger)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing dynatrace metadata: %w", err)
+	}
+
+	httpClient := kedautil.CreateHTTPClient(config.GlobalHTTPTimeout, false)
+
+	logMsg := fmt.Sprintf("Initializing Dynatrace Scaler (Host: %s)", meta.host)
+
+	logger.Info(logMsg)
+
+	return &dynatraceScaler{
+		metricType: metricType,
+		metadata:   meta,
+		httpClient: httpClient,
+		logger:     logger}, nil
+}
+
+func parseDynatraceMetadata(config *scalersconfig.ScalerConfig, logger logr.Logger) (*dynatraceMetadata, error) {
+	meta := dynatraceMetadata{}
+	var err error
+
+	host, err := GetFromAuthOrMeta(config, "host")
+	if err != nil {
+		return nil, err
+	}
+	meta.host = host
+
+	token, err := GetFromAuthOrMeta(config, "token")
```

Comment:
As the token is a secret, we have to get it from metadata or workload environment variables, but we shouldn't read it from metadata